wgsl/index.bs

```diff
@@ -4557,11 +4555,282 @@ TODO: *Stub*: Expression evaluation
 
 ## Uniformity TODO ## {#uniformity}
 
-### Uniform control flow TODO ### {#uniform-control-flow}
+### Terminology and concepts TODO ### {#uniformity-concepts}
+
+The following definitions are merely informative, trying to give an intuition for what the analysis in the next subsection is computing.
+The analysis is what actually defines these concepts, and when a program is valid or break the uniformity rules.
+
+Control-flow is said to be uniform at a point in the program if all invocations execute in lockstep at that program point.
+An expression is said to be uniform if it is in uniform control-flow, and all invocations compute the same value for it.
+A local variable is said to be uniform, if all invocations have the same value in that variable at every program point where it is live.
+
+### Uniformity analysis overview ### {#uniformity-overview}
+
+In this section we specify an analysis that verifies that functions which are only safe to call in uniform control flow (e.g. barriers and derivatives) are only called in such a context.
+
+<div class="note">Note: This analysis has the following desirable properties:
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
+      - Linear time complexity
+      - Refactoring a piece of code into a function cannot make it invalid if it was valid before
+      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used to craft a good error message
+</div>
+
+The analysis analyzes each function in turn, verifying that there is a context where it is safe to call this function. It rejects the program as invalid if there is no such context.
+
+At the same time, it computies metadata about the function to help analyze its callers in turn. This means that the call graph must first be built, and functions must be analyzed from the leaves upwards, i.e. from functions that call no function outside the standard library toward the entry point. This way, whenever a function is analyzed, the metadata for all of its callees has already been computed. There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+Note: another way of saying the same thing is that we do a topological sort of functions ordered by the "is a (possibly indirect) callee of" partial order, and analyze in that order.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the AST of the function, building a directed graph along the way based on the rules in the following subsections.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+<div class="note">Note: apart from two special vertices True and False, all vertices can be understood as having one of the following meanings:
```

Comment:
I am more familiar with referring to locations on a graph as "nodes" which might prevent potential confusion with the other use of "vertices" that is strewn throughout this document.

Reply:
I originally used "nodes", and switched it because I thought that "vertex" the more correct name for this concept. I can very easily rename it back to "node". You have a good point that we already use "vertex" in the spec for completely different things.

Reply:
I agree with @pow2clk .  Vertices are confusing in part because we have "vertex shaders", "vertex attributes", etc. Nodes appear much less ambiguous here.

Reply:
Fixed, thanks